packages/state-machine/src/index.ts

```diff
@@ -0,0 +1,201 @@
+
+/**
+ * Options object for {@link createMachine}.
+ */
+export type MachineOptions<T extends StatesBase<keyof T>> = {
+  states: MachineStates<T>;
+  initial: MachineInitial<T>;
+};
+
+type MachineStateDiscriminator<T extends StatesBase<keyof T>> = {
+  [K in keyof T]: {
+    readonly type: K;
+    readonly value: StateValue<T[K]>;
+    readonly to: MachineNext<T, K>;
+  };
+};
+
+/**
+ * Type of the state object returned by {@link createMachine}.
+ *
+ * Use with a specific state name to get the state object for that state.
+ * @example
+ * ```ts
+ * type IdleState = MachineState<States, "idle">
+ * ```
+ *
+ * Use with `keyof States` to get the union of all state objects.
+ * @example
+ * ```ts
+ * type AllStates = MachineState<States, keyof States>
+ * ```
+ */
+export type MachineState<
+  T extends StatesBase<keyof T>,
+  K extends keyof T,
+> = MachineStateDiscriminator<T>[K];
+
+type PossibleNextKeys<T extends StatesBase<keyof T>, TKey extends keyof T> = Exclude<
+  // @ts-expect-error
+  Extract<keyof T, T[TKey] extends { to: infer To } ? To[number] : any>,
+  TKey | symbol
+>;
+
+/**
+ * Type of the `next` function passed to state functions. Or the `to` property of the state object.
+ *
+ * Use with a specific state name to get the `next` function for that state.
+ * @example
+ * ```ts
+ * type IdleNext = MachineNext<States, "idle">
+ * ```
+ *
+ * Use with `keyof States` to get the union of all `next` functions.
+ * @example
+ * ```ts
+ * type AllNext = MachineNext<States, keyof States>
+ * ```
+ */
+export type MachineNext<T extends StatesBase<keyof T>, TKey extends keyof T> = {
+  readonly [K in PossibleNextKeys<T, TKey>]: (input: StateInput<T[K]>) => void;
+} & (<K extends PossibleNextKeys<T, TKey>>(
+  ...args: T[K] extends { input: infer Input } ? [to: K, input: Input] : [to: K, input?: undefined]
+) => void);
+
+const EQUALS_OPTIONS = { equals: (a: { type: any }, b: { type: any }) => a === b };
+
+/**
+ * Creates a reactive state machine.
+ *
+ * @param options {@link MachineOptions} object.
+ *
+ * @returns A signal with the current state object.
+ *
+ * @example
+ * ```ts
+ * const state = createMachine<{
+ *   idle: { value: "foo"; to: ["loading"] };
+ *   loading: { input: number; value: "bar"; to: ["idle"] };
+ * }>({
+ *   initial: "idle",
+ *   states: {
+ *     idle(input, to) {
+ *       return "foo";
+ *     },
+ *     loading(input, to) {
+ *       setTimeout(() => to("idle"), input);
+ *       return "bar";
+ *     }
+ *   }
+ * })
+ *
+ * state.type // "idle"
+ * state.value // "foo"
+ *
+ * if (state.type === "idle") {
+ *   state.to.loading(1000)
+ *
+ *   state.type // "loading"
+ *   state.value // "bar"
+ * }
+ * ```
+ */
+export function createMachine<T extends StatesBase<keyof T>>(
+  options: MachineOptions<T>,
+): Accessor<MachineState<T, keyof T>> & MachineState<T, keyof T> {
+  const { states, initial } = options;
+
+  const [payload, setPayload] = createSignal(
```

Comment:
Instead of using payload and a memo to access type and value, why not use two signals for type and memo and batch the updates?

Reply:
I don’t see how that would be beneficial. you use more signals to have more fine grain changes, here it’s always meant to be together. a single point of change, even if the value is an object.

Reply:
Storing those values in separate signals saves you the cost of creating objects again and again. Using batch will guarantee the changes will stay in sync.

Reply:
Adding another signal and a batch function would bring it's own complexity, allocations, and issues with managing triggering updates from two separate signals. *(I only want to trigger the memo to rerun if the state changed, not if it was called again with different input)*
And the memo is necessary to achieve the intended behavior.
But I managed to remove one object creation without it.